drivers/network/netio/netio.c

```diff
@@ -0,0 +1,860 @@
+            DbgPrint("WskControlSocket: Option %d Not implemented\n", RequestType);
+    }
+
+err_out:
+    Irp->IoStatus.Status = status;
+    IoCompleteRequest(Irp, IO_NETWORK_INCREMENT);
+
+    return STATUS_PENDING;
+}
+
+static NTSTATUS WSKAPI
+WskCloseSocket(_In_ PWSK_SOCKET Socket, _Inout_ PIRP Irp)
+{
+    NTSTATUS status = STATUS_SUCCESS;
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+
+    IoSetNextIrpStackLocation(Irp);
+
+    SocketPut(s);
+
+    Irp->IoStatus.Status = status;
+    IoCompleteRequest(Irp, IO_NETWORK_INCREMENT);
+
+    return STATUS_PENDING;
+}
+
+static struct _TRANSPORT_ADDRESS *
+TdiTransportAddressFromSocketAddress(PSOCKADDR SocketAddress)
+{
+    struct _TRANSPORT_ADDRESS *ta;
+
+    ta = ExAllocatePoolWithTag(NonPagedPool, sizeof(*ta) + sizeof(struct sockaddr), 'ADDR');
+    if (ta == NULL)
+    {
+        DbgPrint("TdiTransportAddressFromSocketAddress: Out of memory\n");
+        return NULL;
+    }
+
+    ta->TAAddressCount = 1;
+    ta->Address[0].AddressLength = sizeof(SocketAddress->sa_data);
+    ta->Address[0].AddressType = SocketAddress->sa_family;
+    memcpy(&ta->Address[0].Address[0], &SocketAddress->sa_data, ta->Address[0].AddressLength);
+
+    return ta;
+}
+
+static PTDI_CONNECTION_INFORMATION
+TdiConnectionInfoFromSocketAddress(PSOCKADDR SocketAddress)
+{
+    PTRANSPORT_ADDRESS TargetAddress;
+    PTDI_CONNECTION_INFORMATION ConnectionInformation = NULL;
+    NTSTATUS status;
+
+    TargetAddress = TdiTransportAddressFromSocketAddress(SocketAddress);
+    if (TargetAddress == NULL)
+        return NULL;
+
+    status = TdiBuildConnectionInfo(&ConnectionInformation, TargetAddress);
+    ExFreePoolWithTag(TargetAddress, 'ADDR');
+
+    if (!NT_SUCCESS(status))
+    {
+        if (ConnectionInformation != NULL)
+        {
+            ExFreePoolWithTag(ConnectionInformation, TAG_AFD_TDI_CONNECTION_INFORMATION);
+        }
+        return NULL;
+    }
+    return ConnectionInformation;
+}
+
+static NTSTATUS WSKAPI
+WskBind(_In_ PWSK_SOCKET Socket, _In_ PSOCKADDR LocalAddress, _Reserved_ ULONG Flags, _Inout_ PIRP Irp)
+{
+    NTSTATUS status;
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+    PTRANSPORT_ADDRESS ta = TdiTransportAddressFromSocketAddress(LocalAddress);
+
+    IoSetNextIrpStackLocation(Irp);
+
+    if (ta == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        status = STATUS_INSUFFICIENT_RESOURCES;
+        goto err_out;
+    }
+    if (s->LocalAddressHandle != NULL)
+    {
+        ZwClose(s->LocalAddressHandle);
+        s->LocalAddressHandle = NULL;
+        s->LocalAddressFile = NULL;
+    }
+
+    status = TdiOpenAddressFile(&s->TdiName,
+                                ta, AFD_SHARE_REUSE, &s->LocalAddressHandle, &s->LocalAddressFile);
+    if (NT_SUCCESS(status))
+    {
+        memcpy(&s->LocalAddress, LocalAddress, sizeof(s->LocalAddress));
+    }
+    ExFreePoolWithTag(ta, 'ADDR');
```

Comment:
A define for this pool tag plz (and it should be `'RDDA'`).

Reply:
Ok, now we have TAG_NETIO. If I need to use different tags for each allocation please let me know.

Reply:
I think the tags should be ok now.